### DISCUSSION TITLE
Issues with Crazyflie 2.1 Noise in State Estimate

### DISCUSSION THREAD

**Discussion by payamparvizi** [bitcraze/discussions#1621] [Q&A]
I am currently working on implementing reinforcement learning on Crazyflie 2.1, but I am experiencing some issues and would appreciate your help. I should mention that I am not using image processing in my setup; my task is to use RL to command the Crazyflie to hover at a specific altitude.

When I connect the Crazyflie to the Crazyradio on the cfclient platform, I don't receive data for the x and y axes, and the z-axis data is highly noisy. To obtain the Crazyflie's altitude, I use the data['stateEstimate.z'] command. For example, while the Crazyflie is stationary, the altitude reading fluctuates around -75 meters (changes every time when rebooting Crazyflie), with a noise margin of ±0.40 meters, resulting in approximately 40 cm of noise in altitude measurements.

Additionally, I cannot access the "Command-based flight control" section in cfclient. When I issue a takeoff command to reach a 1.0-meter altitude in my code (takeoff(1.0)), the Crazyflie exceeds the intended altitude and hits the ceiling rather than hovering at 1.0 meters.

Based on my research, it seems I might need a Flow Deck to address these issues. Is this correct? Are there alternative components that could help solve this problem? Here are the components I currently have:

- Crazyflie 2.1 (model: CF21KIT)
- Crazy Radio
- AI Deck
- Debug Adapter Kit
- Breakout Expansion Board
- Prototyping Expansion Board

I would be grateful if you could help me resolve these issues.

**Comment by gemenerik** [comment]
To achieve command-based flight control with the Crazyflie 2.1, you’ll need additional positioning sensors. While we may not always enforce this in every example, it’s essential for autonomous flight. The standard Crazyflie only has an IMU and a barometer, which isn’t enough for positioning.

I recommend using a Flow deck for your setup. Despite its downside of relying on dead reckoning for x and y positioning, it should work well for your RL research and it's relatively affordable.
